### PR TITLE
*: Support inspect kv

### PR DESCRIPTION
**Pull request by zimulala**

### CODE REVIEW

File: ddl/stat.go [hunk 1]
```diff
@@ -47,35 +46,12 @@ func (d *ddl) GetScope(status string) variable.ScopeFlag {
...
-		}
-
-		if job != nil {
-			reorgHandle, err = t.GetDDLReorgHandle(job)
-			if err != nil {
-				return errors.Trace(err)
-			}
+		var err1 error
+		info, err1 = inspectkv.GetDDLInfo(txn)
+		if err1 != nil {
```
Comment (siddontang) [inline ddl/stat.go]: return errors.Trace(err1) directly, no need if.

Comment (zimulala) [inline ddl/stat.go]: got it.

File: inspectkv/inspectkv.go [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+	vals := make([]interface{}, len(indexInfo.Columns))
+	cols := table.Cols()
+
+	for i, col := range indexInfo.Columns {
+		key := table.RecordKey(handle, cols[col.Offset])
+		val, err := txn.Get(key)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		vals[i] = val
```
Comment (siddontang) [inline inspectkv/inspectkv.go]: val here is the encoded data, how to get the origin value?

File: inspectkv/inspectkv.go [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+	}
+
+	return vals, nil
+}
+
+// GetIndexHandles returns index handles.
+func GetIndexHandles(table table.Table, txn kv.Transaction, idx *column.IndexedCol) ([]int64, error) {
```
Comment (siddontang) [inline inspectkv/inspectkv.go]: we can't dump whole table index handles if the table is too huge.
must use scan to do this.

File: inspectkv/inspectkv.go [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+		}
+		handles = append(handles, h)
+	}
+
+	return handles, nil
+}
+
+// GetTableData gets table row handles and column values.
+// If there is no special iterator, it will be nil.
+func GetTableData(table table.Table, txn kv.Transaction, it kv.Iterator) ([]int64, []interface{}, error) {
```
Comment (siddontang) [inline inspectkv/inspectkv.go]: don't get whole data for a huge table, use scan.

File: inspectkv/inspectkv.go [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+	}
+	defer snap.Release()
+
+	it, err := snap.Seek([]byte(table.FirstKey()))
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	defer it.Close()
+
+	_, data, err := GetTableData(table, txn, it)
```
Comment (siddontang) [inline inspectkv/inspectkv.go]: the snapshot for the txn is not the same for store.GetSnapshot(ver), why using them together?

File: inspectkv/inspectkv.go (hunk 2 above)
Comment (zimulala) [inline inspectkv/inspectkv.go]: `DecodeValue()`
I will modify it later.

File: inspectkv/inspectkv.go (hunk 3 above)
Comment (zimulala) [inline inspectkv/inspectkv.go]: yep, but maybe in next PR?

File: inspectkv/inspectkv.go (hunk 5 above)
Comment (zimulala) [inline inspectkv/inspectkv.go]: I didn't notice it. I will modify it.

File: inspectkv/inspectkv.go (hunk 3 above)
Comment (siddontang) [inline inspectkv/inspectkv.go]: no, this PR.

File: inspectkv/inspectkv.go (hunk 2 above)
Comment (siddontang) [inline inspectkv/inspectkv.go]: table has a DecodeValue function

Comment (zimulala) [inline inspectkv/inspectkv.go]: yep, I have updated.

File: table/tables/tables.go [hunk 6]
```diff
@@ -528,8 +523,12 @@ func (t *Table) RowWithCols(ctx context.Context, h int64, cols []*column.Col) ([
 // Row implements table.Table Row interface.
 func (t *Table) Row(ctx context.Context, h int64) ([]interface{}, error) {
 	// TODO: we only interested in mentioned cols
-	cols := t.Cols()
-	r, err := t.RowWithCols(ctx, h, cols)
+	txn, err := ctx.GetTxn(false)
```
Comment (siddontang) [inline table/tables/tables.go]: why using txn instead of ctx?

File: table/tables/tables.go [hunk 7]
```diff
@@ -641,17 +640,16 @@ func (t *Table) BuildIndexForRow(rm kv.RetrieverMutator, h int64, vals []interfa
...
-func (t *Table) IterRecords(ctx context.Context, startKey string, cols []*column.Col, fn table.RecordIterFunc) error {
-	txn, err := ctx.GetTxn(false)
-	if err != nil {
-		return errors.Trace(err)
-	}
-
-	it, err := txn.Seek([]byte(startKey))
-	if err != nil {
-		return errors.Trace(err)
+func (t *Table) IterRecords(txn kv.Transaction, it kv.Iterator, startKey string, cols []*column.Col,
```
Comment (siddontang) [inline table/tables/tables.go]: why using txn instead of ctx?

File: table/tables/tables.go (hunk 6 above)
Comment (zimulala) [inline table/tables/tables.go]: inspectkv also invokes `RowWithCols()` but it doesn't have ctx.

File: inspectkv/inspectkv.go (hunk 3 above)
Comment (zimulala) [inline inspectkv/inspectkv.go]: Gotcha.

File: inspectkv/inspectkv.go [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+	info.ReorgHandle, err = t.GetDDLReorgHandle(info.Job)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	return info, nil
+}
+
+// GetIndexVals returns index values.
+func GetIndexVals(table table.Table, txn kv.Transaction, indexInfo model.IndexInfo, handle int64) ([]interface{}, error) {
```
Comment (coocood) [inline inspectkv/inspectkv.go]: What this function is used for?
It just get column values for a row, it doesn't lookup the index at all.

File: inspectkv/inspectkv.go [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+	}
+
+	return handles, nil
+}
+
+// GetTableData gets table row handles and column values.
+// If there is no special iterator, it will be nil.
+func GetTableData(table table.Table, retriever kv.Retriever) ([]int64, []interface{}, error) {
+	var handles []int64
+	var data []interface{}
```
Comment (coocood) [inline inspectkv/inspectkv.go]: The data is a slice of row, a row is a slice of `interface{}`, Maybe change the type to `[][]interface{}` should be better.

File: inspectkv/inspectkv.go (hunk 8 above)
Comment (coocood) [inline inspectkv/inspectkv.go]: The function name is confusing, what does `IndexVals` mean?
the key of an index entry is the value of a column, the value of an index is a handle,
but this function has a handle input argument.

File: inspectkv/inspectkv.go [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+	}
+
+	return vals, nil
+}
+
+// GetIndexHandles returns index handles.
+func GetIndexHandles(table table.Table, txn kv.Transaction, idx *column.IndexedCol) ([]int64, error) {
+	var handles []int64
+	kvIndex := kv.NewKVIndex(table.IndexPrefix(), idx.Name.L, idx.ID, idx.Unique)
+	it, err := kvIndex.SeekFirst(txn)
```
Comment (disksing) [inline inspectkv/inspectkv.go]: The iterator should be closed after use.

File: inspectkv/inspectkv_test.go [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+	c.Assert(handles, HasLen, 2)
+
+	vals, err := GetIndexVals(tb, txn, indices[0].IndexInfo, handles[0])
+	c.Assert(err, IsNil)
+	c.Assert(vals, HasLen, 1)
+	c.Assert(vals[0], Equals, int64(10))
+}
+
+// mockContext represents mocked context.Context.
+type mockContext struct {
```
Comment (siddontang) [inline inspectkv/inspectkv_test.go]: we already has a mockContext, unify them.

**Comment by zimulala** [comment]
got it.

**Comment by zimulala** [comment]
PTAL @siddontang @disksing @coocood

File: inspectkv/inspectkv.go [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+}
+
+func next(data []interface{}) []interface{} {
+	// Add 0x0 to the end of data.
+	return append(data, []byte{})
+}
+
+// ScanIndexData scans the index handles and values in a limited number.
+// It returns data and the next startVals.
+func ScanIndexData(tableIndexPrefix string, idx *column.IndexedCol, txn kv.Transaction,
```
Comment (siddontang) [inline inspectkv/inspectkv.go]: define a struct to put return handle and its associated values together. 

like
```
type IndexRow struct {
 Handle int64
Values []interface{}
}
```

Comment (siddontang) [inline inspectkv/inspectkv.go]: pass KvIndex instead of tableIndexPrefix + idx

Comment (siddontang) [inline inspectkv/inspectkv.go]: put txn as the first argument

File: inspectkv/inspectkv.go [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+	var (
+		count   int
+		handles []int64
+		vals    [][]interface{}
+		curVals []interface{}
+	)
+	for !table.IsLimit(count, limit) {
+		val, h, err1 := it.Next()
+		if terror.ErrorEqual(err1, io.EOF) {
+			return handles, vals, next(curVals), nil
```
Comment (siddontang) [inline inspectkv/inspectkv.go]: return io.EOF directly to tell outer we don't have more data, no need next here.

File: inspectkv/inspectkv.go [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+}
+
+// ScanTableData scans table row handles and column values in a limited number.
+// It returns data and the next startKey.
+func ScanTableData(t table.Table, retriever kv.Retriever, startKey string, limit int) (
+	[]int64, [][]interface{}, string, error) {
+	var handles []int64
+	var data [][]interface{}
+
+	nextKey, err := t.ScanRecords(retriever, startKey, limit, t.Cols(),
```
Comment (siddontang) [inline inspectkv/inspectkv.go]: why not use IterRecord directly and check the limit here?

File: inspectkv/inspectkv.go (hunk 13 above)
Comment (zimulala) [inline inspectkv/inspectkv.go]: but it returns valid data and io.EOF together?

File: inspectkv/inspectkv.go (hunk 12 above)
Comment (zimulala) [inline inspectkv/inspectkv.go]: got it.

File: inspectkv/inspectkv.go (hunk 14 above)
Comment (zimulala) [inline inspectkv/inspectkv.go]: Because IterRecord doesn't return nextKey.

Comment (siddontang) [inline inspectkv/inspectkv.go]: The key for scan table is handle, no need string, and we can use last got handle + 1 for next key.

Comment (zimulala) [inline inspectkv/inspectkv.go]: but the handle only promises not to repeat. So it isn't good to get handle + 1 for next key?

File: table/table.go [hunk 15]
```diff
@@ -110,6 +113,18 @@ type Table interface {
 	LockRow(ctx context.Context, h int64) error
 }
 
+// UnlimitedScanCnt is used for unlimited scanning.
+const UnlimitedScanCnt = -1
+
+// IsLimit returns if count reachs limit.
+func IsLimit(count, limit int) bool {
```
Comment (siddontang) [inline table/table.go]: The function seems unnecessary.
Checking the condition in for loop is not hard, e,g. 

for limit != 0 {
   limit--
}

if limit is -1, we will run max int64 time to get 0 again, and we don't have so huge data.

File: inspectkv/inspectkv.go [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	return info, nil
+}
+
+func end(data []interface{}) []interface{} {
+	// Add 0x0 to the end of data.
+	return append(data, []byte{})
```
Comment (siddontang) [inline inspectkv/inspectkv.go]: use nil, not []byte{}

btw, we should think a better name than "end".

File: sessionctx/variable/session.go [hunk 17]
```diff
@@ -51,7 +51,8 @@ func (k sessionVarsKeyType) String() string {
 	return "session_vars"
 }
 
-const sessionVarsKey sessionVarsKeyType = 0
+// SessionVarsKey is the session variable key.
+const SessionVarsKey sessionVarsKeyType = 0
```
Comment (siddontang) [inline sessionctx/variable/session.go]: why export it?

File: table/table.go [hunk 18]
```diff
@@ -33,11 +33,14 @@ type RecordIterFunc func(h int64, rec []interface{}, cols []*column.Col) (more b
 
 // Table is used to retrieve and modify rows in table.
 type Table interface {
-	// IterRecords iterates records in the table and call fn.
-	IterRecords(ctx context.Context, startKey string, cols []*column.Col, fn RecordIterFunc) error
+	// IterRecords iterates records in the table and calls fn.
+	IterRecords(retriever kv.Retriever, startKey string, cols []*column.Col, fn RecordIterFunc) error
+
+	// ScanRecords scans records in a limited number and calls fn.
```
Comment (coocood) [inline table/table.go]: the returned values should be named and have comment about.

Comment (siddontang) [inline table/table.go]: using IterRecords is enough, maybe we can remove this function.

File: sessionctx/variable/session.go (hunk 17 above)
Comment (zimulala) [inline sessionctx/variable/session.go]: It's used for test.

Comment (siddontang) [inline sessionctx/variable/session.go]: use BindSessionVars instead.

File: inspectkv/inspectkv.go (hunk 16 above)
Comment (zimulala) [inline inspectkv/inspectkv.go]: yep, nil is correct.
Do you have a suggestion?

File: sessionctx/variable/session.go (hunk 17 above)
Comment (zimulala) [inline sessionctx/variable/session.go]: got it.

File: table/table.go (hunk 18 above)
Comment (zimulala) [inline table/table.go]: yep.

File: table/table.go (hunk 15 above)
Comment (zimulala) [inline table/table.go]: good idea.

**Comment by zimulala** [comment]
PTAL @siddontang @coocood @disksing

File: inspectkv/inspectkv.go [hunk 19]
```diff
@@ -0,0 +1,157 @@
...
+}
+
+// ScanTableData scans table row handles and column values in a limited number.
+// It returns data and the next startKey.
+func ScanTableData(t table.Table, retriever kv.Retriever, startHandle, limit int64) (
+	[]int64, [][]interface{}, int64, error) {
+	var handles []int64
+	var data [][]interface{}
+
+	startKey := tables.EncodeRecordKey(t.TableID(), startHandle, 0)
```
Comment (siddontang) [inline inspectkv/inspectkv.go]: use t.RecordKey(startHandle, nil) instead.

File: inspectkv/inspectkv.go [hunk 20]
```diff
@@ -0,0 +1,157 @@
...
+	err := t.IterRecords(retriever, string(startKey), t.Cols(),
```
Comment (siddontang) [inline inspectkv/inspectkv.go]: change using handle for IterRecords next PR or this PR?

**Comment by coocood** [comment]
LGTM

File: inspectkv/inspectkv.go [hunk 21]
```diff
@@ -0,0 +1,156 @@
...
+	} else if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+
+	return idxRows, nextVals, nil
+}
+
+// ScanTableData scans table row handles and column values in a limited number.
+// It returns data and the next startKey.
+func ScanTableData(t table.Table, retriever kv.Retriever, startHandle, limit int64) (
```
Comment (siddontang) [inline inspectkv/inspectkv.go]: same return problem as index, put handle and values together.

File: inspectkv/inspectkv_test.go [hunk 22]
```diff
@@ -0,0 +1,209 @@
...
+	c.Assert(err, IsNil)
+	c.Assert(idxRows, DeepEquals, []*IndexRow{{Handle: int64(2), Values: []interface{}{int64(20)}}})
+	idxRows, nextVals, err = ScanIndexData(txn, kvIndex, nextVals, 1)
+	c.Assert(idxRows, IsNil)
+	c.Assert(nextVals, DeepEquals, []interface{}{nil})
+	c.Assert(err, IsNil)
+}
+
+// mockContext represents mocked context.Context.
+type mockContext struct {
```
Comment (siddontang) [inline inspectkv/inspectkv_test.go]: unify this with util/mock/context

File: ddl/column_test.go [hunk 23]
```diff
@@ -121,7 +121,9 @@ func (s *testColumnSuite) TestColumn(c *C) {
 	c.Assert(err, IsNil)
 
 	i := int64(0)
-	t.IterRecords(ctx, t.FirstKey(), t.Cols(), func(h int64, data []interface{}, cols []*column.Col) (bool, error) {
+	txn, err := ctx.GetTxn(false)
```
Comment (qiuyesuifeng) [inline ddl/column_test.go]: U can use `ctx.GetTxn(true)`, it inner call `FinishTxn `, then u can remove line 120 to 121.

File: ddl/column_test.go [hunk 24]
```diff
@@ -156,7 +160,9 @@ func (s *testColumnSuite) TestColumn(c *C) {
 	err = ctx.FinishTxn(false)
 	c.Assert(err, IsNil)
 
-	values, err := t.RowWithCols(ctx, h, t.Cols())
+	txn, err = ctx.GetTxn(false)
```
Comment (qiuyesuifeng) [inline ddl/column_test.go]: Seems needless, in your code here u have not commit the txn, so u can directly use txn in line 145.

File: inspectkv/inspectkv.go [hunk 25]
```diff
@@ -0,0 +1,163 @@
...
+	info.Job, err = t.GetDDLJob(0)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	info.SchemaVer, err = t.GetSchemaVersion()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if info.Job == nil {
+		return info, nil
```
Comment (qiuyesuifeng) [inline inspectkv/inspectkv.go]: Why not directly return `nil, nil`?

File: inspectkv/inspectkv.go [hunk 26]
```diff
@@ -0,0 +1,163 @@
...
+
+	info.ReorgHandle, err = t.GetDDLReorgHandle(info.Job)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	return info, nil
+}
+
+func end(data []interface{}) []interface{} {
```
Comment (qiuyesuifeng) [inline inspectkv/inspectkv.go]: What is the meaning of end?
Seems confused.

File: inspectkv/inspectkv.go [hunk 27]
```diff
@@ -0,0 +1,163 @@
...
+}
+
+// IndexRow is the index information composed of a handle and values.
+type IndexRow struct {
+	Handle int64
+	Values []interface{}
+}
+
+// ScanIndexData scans the index handles and values in a limited number.
+// It returns data and the next startVals.
```
Comment (qiuyesuifeng) [inline inspectkv/inspectkv.go]: Has seen the `data and the next startVals`.

File: inspectkv/inspectkv.go [hunk 28]
```diff
@@ -0,0 +1,163 @@
...
+// ScanTableData scans table row handles and column values in a limited number.
+// It returns data and the next startKey.
+func ScanTableData(t table.Table, retriever kv.Retriever, startHandle, limit int64) (
+	[]*RecordData, int64, error) {
+	var records []*RecordData
+
+	startKey := t.RecordKey(startHandle, nil)
+	err := t.IterRecords(retriever, string(startKey), t.Cols(),
+		func(h int64, d []interface{}, cols []*column.Col) (bool, error) {
+			for limit != 0 {
```
Comment (qiuyesuifeng) [inline inspectkv/inspectkv.go]: I see u directly return in for clause, so use if is better?

Comment (siddontang) [inline inspectkv/inspectkv.go]: seem must use if

File: inspectkv/inspectkv.go (hunk 26 above)
Comment (siddontang) [inline inspectkv/inspectkv.go]: maybe nextIndexVals?

File: inspectkv/inspectkv_test.go [hunk 29]
```diff
@@ -0,0 +1,161 @@
...
+	record.Values = []interface{}{int64(20)}
+	c.Assert(records, DeepEquals, []*RecordData{record})
+	startHandle := nextHandle
+	records, nextHandle, err = ScanTableData(tb, txn, startHandle, 1)
+	c.Assert(records, IsNil)
+	c.Assert(nextHandle, Equals, startHandle)
+	c.Assert(err, IsNil)
+
+	kvIndex := kv.NewKVIndex(tb.IndexPrefix(), indices[0].Name.L, indices[0].ID, indices[0].Unique)
+	idxRows, nextVals, err := ScanIndexData(txn, kvIndex, []interface{}{int64(10)}, 1)
```
Comment (siddontang) [inline inspectkv/inspectkv_test.go]: use a limit > 1, like 2  to test too.

File: util/mock/context.go [hunk 30]
```diff
@@ -49,12 +51,30 @@ func (c *Context) ClearValue(key fmt.Stringer) {
...
+	var err error
+	c.txn, err = c.Store.Begin()
+
+	return c.txn, err
 }
 
 // FinishTxn implements context.Context FinishTxn interface.
 func (c *Context) FinishTxn(rollback bool) error {
-	return nil
+	if c.Store == nil || c.txn == nil {
```
Comment (siddontang) [inline util/mock/context.go]: no need to check c.Store == nil here, if store is nil, the txn must always be nil.

File: ddl/column_test.go (hunk 23 above)
Comment (zimulala) [inline ddl/column_test.go]: got it.

File: inspectkv/inspectkv.go (hunk 26 above)
Comment (zimulala) [inline inspectkv/inspectkv.go]: nextIndexVals is better, I will modify it.

File: util/mock/context.go (hunk 30 above)
Comment (zimulala) [inline util/mock/context.go]: yep.

File: inspectkv/inspectkv.go (hunk 28 above)
Comment (zimulala) [inline inspectkv/inspectkv.go]: yep, I will modify it.

File: inspectkv/inspectkv_test.go [hunk 31]
```diff
@@ -0,0 +1,161 @@
...
+		Type:     model.ActionCreateSchema,
+	}
+	err = t.EnQueueDDLJob(job)
+	c.Assert(err, IsNil)
+	info, err := GetDDLInfo(txn)
+	c.Assert(err, IsNil)
+	c.Assert(info.Owner, DeepEquals, owner)
+	c.Assert(info.Job, DeepEquals, job)
+	c.Assert(info.ReorgHandle, Equals, int64(0))
+	txn.Commit()
```
Comment (qiuyesuifeng) [inline inspectkv/inspectkv_test.go]: Check error

File: inspectkv/inspectkv.go [hunk 32]
```diff
@@ -0,0 +1,163 @@
...
+			}
+
+			return false, nil
+		})
+	if err != nil {
+		return nil, 0, errors.Trace(err)
+	}
+
+	if len(records) == 0 {
+		return records, startHandle, nil
```
Comment (qiuyesuifeng) [inline inspectkv/inspectkv.go]: Seems not the same with ScanIndexData?

File: inspectkv/inspectkv.go (hunk 25 above)
Comment (zimulala) [inline inspectkv/inspectkv.go]: consistent with the original code.

File: inspectkv/inspectkv_test.go (hunk 29 above)
Comment (zimulala) [inline inspectkv/inspectkv_test.go]: got it.

File: inspectkv/inspectkv_test.go (hunk 31 above)
Comment (zimulala) [inline inspectkv/inspectkv_test.go]: got it.

File: inspectkv/inspectkv.go (hunk 32 above)
Comment (zimulala) [inline inspectkv/inspectkv.go]: here `len(records) == 0`, so startHanle is last handle.

**Comment by siddontang** [comment]
LGTM

**Pull request merged**